source/extensions/filters/http/admission_control/admission_control.h

```diff
@@ -0,0 +1,222 @@
+  virtual ~ThreadLocalControllerImpl() = default;
+  virtual void recordSuccess() override { recordRequest(true); }
+  virtual void recordFailure() override { recordRequest(false); }
+
+  virtual uint32_t requestTotalCount() override {
+    maybeUpdateHistoricalData();
+    return global_data_.requests;
+  }
+  virtual uint32_t requestSuccessCount() override {
+    maybeUpdateHistoricalData();
+    return global_data_.successes;
+  }
+
+private:
+  struct RequestData {
+    RequestData() : requests(0), successes(0) {}
+    uint32_t requests;
+    uint32_t successes;
+  };
+
+  void recordRequest(const bool success);
+
+  // Potentially remove any stale samples and record sample aggregates to the historical data.
+  void maybeUpdateHistoricalData();
+
+  // Returns the age of the oldest sample in the historical data.
+  std::chrono::microseconds ageOfOldestSample() const {
+    ASSERT(!historical_data_.empty());
+    using namespace std::chrono;
+    return duration_cast<microseconds>(time_source_.monotonicTime() -
+                                       historical_data_.front().first);
+  }
+
+  // Returns the age of the newest sample in the historical data.
+  std::chrono::microseconds ageOfNewestSample() const {
+    ASSERT(!historical_data_.empty());
+    using namespace std::chrono;
+    return duration_cast<microseconds>(time_source_.monotonicTime() -
+                                       historical_data_.back().first);
+  }
+
+  // Removes the oldest sample in the historical data and reconciles the global data.
+  void removeOldestSample() {
+    ASSERT(!historical_data_.empty());
+    global_data_.successes -= historical_data_.front().second.successes;
+    global_data_.requests -= historical_data_.front().second.requests;
+    historical_data_.pop_front();
+  }
+
+  TimeSource& time_source_;
+  std::deque<std::pair<MonotonicTime, RequestData>> historical_data_;
+
+  // Request data aggregated for the whole look-back window.
+  RequestData global_data_;
+
+  // The rolling time window size.
+  std::chrono::seconds sampling_window_;
+};
+
+using AdmissionControlProto =
+    envoy::extensions::filters::http::admission_control::v3alpha::AdmissionControl;
+
+/**
+ * Determines of a request was successful based on response headers.
+ */
+class ResponseEvaluator {
+public:
+  virtual ~ResponseEvaluator() = default;
+  virtual bool isHttpSuccess(uint64_t code) const PURE;
+  virtual bool isGrpcSuccess(Grpc::Status::GrpcStatus status) const PURE;
+};
+
+class DefaultResponseEvaluator : public ResponseEvaluator {
+public:
+  DefaultResponseEvaluator(AdmissionControlProto::DefaultSuccessCriteria success_criteria);
+  virtual bool isHttpSuccess(uint64_t code) const override;
+  virtual bool isGrpcSuccess(Grpc::Status::GrpcStatus status) const override;
+
+private:
+  std::vector<std::function<bool(uint64_t)>> http_success_fns_;
+  std::unordered_set<uint64_t> grpc_success_codes_;
```

Comment:
It just occurred to me that there are a handful of gRPC success codes. It'll be more performant to just do a sequential search over a vector instead of hashing in an unordered set.

I'll make that change next round.